File: tutorials/matmul.cc
//! @h1 C++ DSL API tutorial: Matrix Multiplication
//! This tutorial will guide you through the basic usage of the C++ DSL API.

#include <gtest/gtest.h>

#include "cinn/cinn.h"

using namespace cinn;  // NOLINT

//! @IGNORE-NEXT
TEST(matmul, basic) {
  //! @h2 Computation basic defition
  //! Declare some varialbe for latter usages.
  Expr M(100), N(200), K(50);
  Var k(K, "k0");  // the reduce axis

  //! Placeholder represents the input arguments for a computation.
  Placeholder<float> A("A", {M, K});
  Placeholder<float> B("B", {K, N});

  //! Define a computation to get the result tensor C.
  auto C =
      Compute({M, N} /*domain*/, [=](Expr i, Expr j) { return ReduceSum(A(i, k) * B(k, j), {k} /*reduce axis*/); });

  //! Generate the stages to get the default schedules.
  auto stages = CreateStages({C} /*the endpoints*/);

  //! Print the generated IR, the `Lower` method just map a computation to the underlying CINN IR.
  auto fn = Lower("fn0", stages, {A, B, C} /*argument list of the fn*/);
  LOG(INFO) << "fn0:\n" << fn;
  //! This will generate the code like
  //! @ROC[c++]
  auto target_source = R"ROC(
function fn0 (_A, _B, _tensor)
{
  for (i, 0, 100)
  {
    for (j, 0, 200)
    {
      tensor__reduce_init[i, j] = 0
    }
  }
  for (i, 0, 100)
  {
    for (j, 0, 200)
    {
      for (k0, 0, 50)
      {
        tensor[i, j] = (tensor[i, j] + (A[i, k0] * B[k0, j]))
      }
    }
  }
}
)ROC";

  //! @IGNORE-NEXT
  ASSERT_EQ(utils::GetStreamCnt(fn), utils::Trim(target_source));

  //! Print the IR as C code
  Target x;
  CodeGenC codegen(x);
  Module::Builder builder("module0", Target());
  builder.AddFunction(fn);
  codegen.SetInlineBuiltinCodes(false);  // Disable inserting the predefined runtime codes to the generated code.
  std::string C_source = codegen.Compile(builder.Build(), CodeGenC::OutputKind::CImpl);
  //! @IGNORE-NEXT
  LOG(INFO) << "C:\n" << C_source;
  //! and will get some code like
  //! @ROC[c++]
  target_source = R"ROC(
#include <cinn_runtime.h>
#include <stdio.h>

void fn0(void* _args, int32_t num_args)
{
  const cinn_buffer_t* _A = cinn_pod_value_to_buffer_p(&(((cinn_pod_value_t*)(_args))[0]));
  const cinn_buffer_t* _B = cinn_pod_value_to_buffer_p(&(((cinn_pod_value_t*)(_args))[1]));
  cinn_buffer_t* _tensor = cinn_pod_value_to_buffer_p(&(((cinn_pod_value_t*)(_args))[2]));
  cinn_buffer_malloc((void*)(0), _tensor);
  const float* A = ((const float*)(_A->memory));
  const float* B = ((const float*)(_B->memory));
  float* tensor = ((float*)(_tensor->memory));
  float* tensor__reduce_init = ((float*)(_tensor->memory));
  for (int32_t i = 0; i < 100; i += 1) {
    for (int32_t j = 0; j < 200; j += 1) {
      tensor__reduce_init[((200 * i) + j)] = 0;
    }
  }
  for (int32_t i = 0; i < 100; i += 1) {
    for (int32_t j = 0; j < 200; j += 1) {
      for (int32_t k0 = 0; k0 < 50; k0 += 1) {
        tensor[((200 * i) + j)] = (tensor[((200 * i) + j)] + (A[((50 * i) + k0)] * B[((200 * k0) + j)]));
      }
    }
  }
  cinn_buffer_free((void*)(0), _tensor);
}
  )ROC";

  //! @IGNORE-NEXT
  ASSERT_EQ(utils::Trim(C_source), utils::Trim(target_source));

  //! @h2 Basic schedule
  //! The computation defines the basic way to compute the result while the schedules will guide the system to generate
  //! different codes. Each kind of code will result in different performance.

  //! Lets create a new stages to hold some schedules.
  auto stages1 = CreateStages({C});

  //! `Tile` method will split the 0-th and 1-th axis tile by tile of 4.
  stages1[C]->Tile(0, 1, 4, 4);

  //! The newly generated code is as follows
  //! @IGNORE-NEXT
  auto fn1 = Lower("fn1", stages1, {A, B, C});
  //! @IGNORE-NEXT
  LOG(INFO) << "fn1:\n" << fn1;

  //! @ROC[c++]
  target_source = R"ROC(
function fn1 (_A, _B, _tensor)
{
  for (i, 0, 100)
  {
    for (j, 0, 200)
    {
      tensor__reduce_init[i, j] = 0
    }
  }
  for (i_outer, 0, 25)
  {
    for (i_inner, 0, 4)
    {
      for (j_outer, 0, 50)
      {
        for (j_inner, 0, 4)
        {
          for (k0, 0, 50)
          {
            tensor[((4 * i_outer) + i_inner), ((4 * j_outer) + j_inner)] = (tensor[((4 * i_outer) + i_inner), ((4 * j_outer) + j_inner)] + (A[((4 * i_outer) + i_inner), k0] * B[k0, ((4 * j_outer) + j_inner)]))
          }
        }
      }
    }
  }
})ROC";

  //! @IGNORE-NEXT
  ASSERT_EQ(utils::GetStreamCnt(fn1), utils::Trim(target_source));
}

File: tutorials/matmul.py
# Copyright (c) 2021 CINN Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
Ways to optimize Matrix Multiplication on CPU
=========================================================

In this tutorial, we will introduce several ways to optimize the performance of the matrix multiplication on X86 CPU.
"""

import cinn
import numpy as np
import time
from cinn import runtime
# sphinx_gallery_thumbnail_path = './paddlepaddle.png'

##################################################################
# Declare the basic computation for a matmul
m = cinn.Expr(1024)
n = cinn.Expr(1024)
k = cinn.Expr(1024)

A = cinn.Placeholder("float32", "A", [m, k])
B = cinn.Placeholder("float32", "B", [k, n])

# k1 is a reduce axis
k1 = cinn.Var(k.as_int32(), "k1")

C = cinn.compute([
    m, n
], lambda vs: cinn.reduce_sum(A(vs[0], k1.expr()) * B(k1.expr(), vs[1]), [k1]),
                 "C")

stages = cinn.create_stages([C])

##################################################################
# Fake input data, here we create a runtime buffer for each input of the generated function.
a = runtime.cinn_buffer_t(
    np.random.randn(m.int(), k.int()).astype("float32"),
    runtime.cinn_x86_device, 32)
b = runtime.cinn_buffer_t(
    np.random.randn(m.int(), k.int()).astype("float32"),
    runtime.cinn_x86_device, 32)
c = runtime.cinn_buffer_t(
    np.zeros([m.int(), n.int()]).astype("float32"), runtime.cinn_x86_device,
    32)


##################################################################
# Here is a helper function to JIT compile the generated program and test the performance
def test_performance(stages,
                     fn_inputs=[A.to_tensor(), B.to_tensor(), C],
                     input_args=[a, b, c]):
    '''
    fake input data, compile and test program's performance
    '''
    target = cinn.Target()
    builder = cinn.Module.Builder("matmul", target)

    func = cinn.lower("matmul", stages, fn_inputs)
    builder.add_function(func)
    module = builder.build()

    jit = cinn.ExecutionEngine()
    jit.link(module)

    args = [runtime.cinn_pod_value_t(_) for _ in input_args]
    matmul = jit.lookup("matmul")

    repeat = 4

    tic = time.perf_counter()
    for i in range(repeat):
        matmul(args)
    toc = time.perf_counter()
    miniseconds = (toc - tic) / repeat * 1e3
    print(f"Takes {miniseconds:0.3f} ms")


# The basic computation without any schedule has a performance as follows
test_performance(stages)

##################################################################
# Blocking
# ---------------------
stages = cinn.create_stages([C])
bn = 32
i_outer, i_inner, j_outer, j_inner = stages[C].tile(0, 1, bn, bn)
k_outer, k_inner = stages[C].split("k1", 4)
stages[C].reorder([i_outer, j_outer, k_outer, k_inner, i_inner, j_inner])

# The performance is
test_performance(stages)

##################################################################
# Vectorization
# ---------------------
stages = cinn.create_stages([C])
bn = 32
i_outer, i_inner, j_outer, j_inner = stages[C].tile(0, 1, bn, bn)
k_outer, k_inner = stages[C].split("k1", 4)
stages[C].reorder([i_outer, j_outer, k_outer, k_inner, i_inner, j_inner])
stages[C].vectorize(j_inner, 8)

# The performance is
test_performance(stages)

##################################################################
# Loop Permutation
# ---------------------
stages = cinn.create_stages([C])
i_outer, i_inner, j_outer, j_inner = stages[C].tile(0, 1, bn, bn)
k_outer, k_inner = stages[C].split("k1", 4)
stages[C].reorder([i_outer, j_outer, k_outer, i_inner, k_inner, j_inner])
stages[C].vectorize(j_inner, 8)
stages[C].unroll(5)

test_performance(stages)

##################################################################
# Array Packing
# ---------------------
packedB = cinn.compute(
    [n / bn, k, cinn.Expr(bn)], lambda x: B(x[1], x[0] * bn + x[2]), "packedB")
C = cinn.compute([m, n], lambda x: cinn.reduce_sum(
    A(x[0], k1.expr()) * packedB(x[1] / bn, k1.expr(), x[1] % bn), [k1]), "C")

stages = cinn.create_stages([C])
stages[packedB].vectorize(2, 8)

i_outer, i_inner, j_outer, j_inner = stages[C].tile(0, 1, bn, bn)
k_outer, k_inner = stages[C].split("k1", 4)
stages[C].reorder([i_outer, j_outer, k_outer, i_inner, k_inner, j_inner])
stages[C].vectorize(j_inner, 8)

# We make the packedB as another input of the generated function and allocate a runtime buffer for it.
packedB_buf = runtime.cinn_buffer_t(
    np.zeros([n.int() // bn, k.int(), bn]).astype("float32"),
    runtime.cinn_x86_device, 32)

# The final performance is
test_performance(
    stages,
    fn_inputs=[A.to_tensor(), B.to_tensor(), C, packedB],
    input_args=[a, b, c, packedB_buf])
